### PR TITLE
pytorch seq2seq implementation for Thai romanization

### PR DESCRIPTION
**Pull request by artificiala**
This commit makes changes in the implementation of seq2seq model as follows:

- add attention mechanism to the seq2seq model
- encapsulate `Encoder` and `Decoder` into 1 class `Seq2Seq`
- adjust the code to let the model train in mini-batches

The name of this model is `thai2rom-pytorch-attn`. The download link and information of the model was added to this file: https://github.com/PyThaiNLP/pythainlp-corpus/blob/master/db.json

------

__Evaluation result__

The model was evaluated with 3 metrics including F1-score, Exact match, Exact match at character level on the validation set (20% of the dataset or 129,642 examples). 

- F1 (macro-average):  0.987
- Exact match:  0.883
- Exact match (Character-level): 0.949

__Examples of model prediction can be seen below:__

```
Example: 1
      input: ขัตติยา กรีมี
groundtruth: khattiya krimi
 prediction: khattiya krimi

Example: 2
      input: คมสรร บุญทอง
groundtruth: khomsan bunthong
 prediction: khomsan bunthong

Example: 3
      input: ต่อเลขหมาย
groundtruth: tolekmai
 prediction: tolekmai

Example: 4
      input: กิตติศักดิ์ หวังวรวงศ์
groundtruth: kittisak wangworawong
 prediction: kittisak wangworawong

Example: 5
      input: สถาปัตยกรรมศาสตร์
groundtruth: sathapattayakamsat
 prediction: sathapattayakammasat

Example: 6
      input: กิมเจียว แซ่โง้ว
groundtruth: kimchiao sae-ngow
 prediction: kimchiao sae-ngow

Example: 7
      input: เขยื้อน รังสิยีรานนท์
groundtruth: khayuean rangsiyiranon
 prediction: khayuean rangsiyiranon

Example: 8
      input: ขนิษฐา เสมอใจ
groundtruth: khanittha samoechai
 prediction: khanittha samoechai

Example: 9
      input: กนกกุล หวังบุญผาติ
groundtruth: kanokkun wangbunphati
 prediction: kanokkun wangbunphat

Example: 10
      input: จรัล กิติสาย
groundtruth: charan kitisai
 prediction: charan kitisai

Example: 11
      input: โกศัย มนัสศิลานนท์
groundtruth: kosai manatsilanon
 prediction: kosai manatsilanon

Example: 12
      input: เกศินี บำเพ็ญนรกิจ
groundtruth: kesini bamphennorakit
 prediction: kesini bamphennonkit

Example: 13
      input: จำรัส จันทร์อำพร
groundtruth: chamrat chan-amphon
 prediction: chamrat chan-amphon

Example: 14
      input: กิติสร สุขประดิษฐ์
groundtruth: kitison sukpradit
 prediction: kitison sukpradit

Example: 15
      input: จุฑามาศ ศรีโกมุท
groundtruth: chuthamat sikomut
 prediction: chuthamat sikomut

Example: 16
      input: กัลยา ผาสุขดี
groundtruth: kanlaya phasukdi
 prediction: kanlaya phasukdi

```

__Calling `romanize` method from `pythainlp.transliterate`:__

```python
>>> from pythainlp.transliterate import romanize
>>> romanize('แมว', engine='thai2rom')
'maeo'
>>> romanize('บ้านไร่', engine='thai2rom')
'banrai'
>>> romanize('บางกอก', engine='thai2rom')
'bangkok'
>>> romanize('อำเภอ', engine='thai2rom')
'amphoe'
```

### CODE REVIEW

File: pythainlp/transliterate/thai2rom.py [hunk 1]
```diff
@@ -50,102 +69,276 @@ def _prepare_sequence_in(self, input_text):
...
+        # sequences_lengths: (batch_size)
+
+        batch_size = sequences.size(0)
+        self.hidden = self.init_hidden(batch_size)
+
+        sequences_lengths = np.sort(sequences_lengths)[::-1]
+        index_sorted = np.argsort(
+            -sequences_lengths
+        )  # use negation in sort in descending order
+        index_unsort = np.argsort(index_sorted)  # to unsorted sequence
```
Comment (heytitle) [inline pythainlp/transliterate/thai2rom.py]: Hi,

May I ask you what you're trying to do in these two lines above?

File: pythainlp/transliterate/thai2rom.py [hunk 2]
```diff
@@ -50,102 +69,276 @@ def _prepare_sequence_in(self, input_text):
...
+            bidirectional=True,
+            batch_first=True,
+        )
+
+        self.dropout = nn.Dropout(dropout)
+
+    def forward(self, sequences, sequences_lengths):
+
+        # sequences: (batch_size, sequence_length=MAX_LENGTH)
+        # sequences_lengths: (batch_size)
```
Comment (heytitle) [inline pythainlp/transliterate/thai2rom.py]: It would be better to remove these comments.

File: pythainlp/transliterate/thai2rom.py [hunk 3]
```diff
@@ -50,102 +69,276 @@ def _prepare_sequence_in(self, input_text):
...
+        mask = source_seq != self.pad_idx
+        return mask
+
+    def forward(
+        self, source_seq, source_seq_len, target_seq, teacher_forcing_ratio=0.5
+    ):
+
+        # source_seq: (batch_size, MAX_LENGTH)
+        # source_seq_len: (batch_size, 1)
+        # target_seq: (batch_size, MAX_LENGTH)
```
Comment (heytitle) [inline pythainlp/transliterate/thai2rom.py]: It would be better to remove these comments.

File: tests/test_tokenize.py [hunk 4]
```diff
@@ -202,10 +202,17 @@ def test_syllable_tokenize(self):
...
+        # in order to test this branch `thai-romanize-pytorch`
+        # the following test should be omitted,
+        # when `pythainlp.tokenize.syllable_tokenize` is implemented
+        # (uncomment the following section)
+        # ------------------------------------------------#
+        #
+        # self.assertEqual(
+        #     syllable_tokenize("แมวกินปลา", engine="ssg"),
+        #     ['แมว', 'กิน', 'ปลา']
+        # )
```
Comment (heytitle) [inline tests/test_tokenize.py]: I guess this test isn't relevant for this PR. Should we remove it or move it to a corresponding PR?

File: pythainlp/transliterate/thai2rom.py [hunk 5]
```diff
@@ -50,102 +69,276 @@ def _prepare_sequence_in(self, input_text):
...
+                                             None, 0)
 
-        return "".join(decoded_seq[:-1])
+        try:
+            target_tensor = (
+                torch.argmax(target_tensor_logits.squeeze(1), 1).cpu().numpy()
+            )
+            target_indices = [t for t in target_tensor]
+            target = [self._ix_to_target_char[t] for t in target_tensor]
+        except Exception as e:
```
Comment (heytitle) [inline pythainlp/transliterate/thai2rom.py]: Should we use a more specific exception?

File: pythainlp/transliterate/thai2rom.py (hunk 1 above)
Comment (artificiala) [reply pythainlp/transliterate/thai2rom.py]: The two lines above are to initialize the hidden state and cell state of the first time step `(h_0, c_0)` to zero. 

However, I think that those two lines and the method `init_hiden` could be removed because if (h_0, c_0) is not provided, both h_0 and c_0 default to zero according to the PyTorch documentation.

File: pythainlp/transliterate/thai2rom.py (hunk 5 above)
Comment (artificiala) [reply pythainlp/transliterate/thai2rom.py]: I have changed from try-catch to if statement, to check the output tensor dimension from Seq2Seq model.

Firstly, it is to handle the case when the `torch.argmax` raise an exception.
This due to the following code segment within the method named `forward` of Seq2seq class return an empty tensor 

```
 for di in range(max_len):
     ...
     if inference and decoder_input == end_token:
                return outputs[:di]
    ...
```

if the variable `di` is 0 (first iteration), and the first decoder output
is the end token <END>, it would return outputs[:0] which is an
empty Tensor, torch.Size([0]).

**Comment by wannaphongcom** [comment]
Colab for testing : https://colab.research.google.com/drive/1w1GpJt5HECaBtVJcEcKL6Xk24lxZjdUb

**Comment by wannaphongcom** [comment]
Now, Can i merge this pull request?

**Comment by artificiala** [comment]
> Now, Can i merge this pull request?

Yes, this pull request is now ready.
